### PR TITLE
update Kubernetes docs for operations

### PR DESCRIPTION
**Pull request by izakp**
Update docs with some notes about Kubernetes operations
- Using `kubectl` to interact with clusters
- Sshing into a node and debugging with system tools
- Draining and replacing problematic nodes

### CODE REVIEW

File: playbooks/kubernetes.md [hunk 1]
```diff
@@ -71,8 +85,35 @@ To diagnose application failures, navigate to the Kubernetes dashboard and searc
...
-result in brief downtime as it will delete _all_ running pods simultaneously.
+command `hokusai [staging|production] refresh`
+
+#### When things go wrong (cluster level)
+
+Kubernetes cluster nodes may experience issues with their underlying system (too many open file handles, full disk, memory, etc) and require system-level debugging, or occasionally need to be re-provisioned.
+
+To inspect a node, first find its FQDN, which is the private DNS name assigned by AWS EC2 autoscaling.
+
+To view all nodes for a given cluster, run `kubectl get nodes`.  Another useful command, if you wish to see pods along with the node on which they are scheduled / running is `kubectl get pods -o wide`
```
Comment (dleve123) [inline playbooks/kubernetes.md]: You can also get the the FQDN of a pod's node via `hokusai [staging|production] status`, for example:

```
dl@phoenix:~/src/force (healthier-healthchecks)
$ hokusai staging status

Resources
===========
NAME                              DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE       CONTAINERS   IMAGES                                                                                        SELECTOR
deployment.extensions/force-web   3         3         3            3           91d       force-web    585031190124.dkr.ecr.us-east-1.amazonaws.com/force:6b1a62fc782b61d456f92e72225f53561c7ef2cc   app=force,component=web,layer=application

NAME                TYPE           CLUSTER-IP      EXTERNAL-IP                                                               PORT(S)                      AGE       SELECTOR
service/force-web   LoadBalancer   100.68.109.20   a099d3c23bb1e11e891cf1273498b4c6-1023438724.us-east-1.elb.amazonaws.com   443:31693/TCP,80:30170/TCP   91d       app=force,component=web,layer=application

Pods
===========
NAME                         READY     STATUS    RESTARTS   AGE       IP               NODE
force-web-749f46cd79-c2tsv   1/1       Running   0          10h       100.123.47.36    ip-10-0-1-235.ec2.internal
force-web-749f46cd79-gf5m7   1/1       Running   0          11h       100.120.48.176   ip-10-0-2-161.ec2.internal
force-web-749f46cd79-t6p82   1/1       Running   0          11h       100.102.237.62   ip-10-0-1-233.ec2.internal
```

Might be a bit more useful if trying to quickly triangulate which hosts could be problematic for a given service. Food for thought!

File: playbooks/kubernetes.md [hunk 2]
```diff
@@ -71,8 +85,35 @@ To diagnose application failures, navigate to the Kubernetes dashboard and searc
...
+
+We pre-install useful packages for system level debugging in our base AMI for Kubernetes machines, so you have `htop` / `lsof` / `netstat` / `tcpdump` all available.  Another useful tool that aims to provide debugging for container environments is [`sysdig`](https://github.com/draios/sysdig/wiki/sysdig-user-guide) and its curses-interface [`csysdig`](https://github.com/draios/sysdig/wiki/Csysdig-Overview).  With it you can get stacktraces, inspect file allocations, network connections, etc while filtering on a specific process or container name.
+
+If a node is for any reason unable to schedule containers, serve traffic or unresponsive, it can safely be terminated - AWS autoscaling will replace the machine automatically with a new one and it will rejoin the cluster.  However, terminating a node that is running application pods may result in service disruption for that application if they are the _only_ pods that back that service.
```
Comment (dleve123) [inline playbooks/kubernetes.md]: I'd recommend mentioning `hokusai [staging|production] status` here as it's a good way to answer the question, "do my service's pods only exist on this host". Also, IIRC, assuming > 1 pods for a service, K8s should distribute across many hosts, correct? Would it be worth mentioning that here?

File: playbooks/kubernetes.md [hunk 3]
```diff
@@ -71,8 +85,35 @@ To diagnose application failures, navigate to the Kubernetes dashboard and searc
...
+
+To evict all running pods from a node before terminating it, use the following procedure:
+
+1) Check all the pods running on the node with `kubectl get pods -o wide --all-namespaces | grep {node-FQDN}`
+
+2) If you are concerned that forcefully evicting any of those pods would result in service disruption, scale up the relevant deployments with `kubectl scale deployment/{deployment-name} --replicas {desired-replicas}`
```
Comment (dleve123) [inline playbooks/kubernetes.md]: I believe this is only a temporary scale, right? The next `hokusai [staging|production] update` with a different number of desired replicas will override this? If so, I'd briefly call this out.

File: playbooks/kubernetes.md [hunk 4]
```diff
@@ -71,8 +85,35 @@ To diagnose application failures, navigate to the Kubernetes dashboard and searc
...
+
+We pre-install useful packages for system level debugging in our base AMI for Kubernetes machines, so you have `htop` / `lsof` / `netstat` / `tcpdump` all available.  Another useful tool that aims to provide debugging for container environments is [`sysdig`](https://github.com/draios/sysdig/wiki/sysdig-user-guide) and its curses-interface [`csysdig`](https://github.com/draios/sysdig/wiki/Csysdig-Overview).  With it you can get stacktraces, inspect file allocations, network connections, etc while filtering on a specific process or container name.
+
+If a node is for any reason unable to schedule containers, serve traffic or unresponsive, it can safely be terminated - AWS autoscaling will replace the machine automatically with a new one and it will rejoin the cluster.  However, terminating a node that is running application pods may result in service disruption for that application if they are the _only_ pods that back that service.
+
+To evict all running pods from a node before terminating it, use the following procedure:
```
Comment (dleve123) [inline playbooks/kubernetes.md]: Do we want to point developers to the [existing `rotate_cluster_nodes`](https://github.com/artsy/substance/blob/c7dee522551e764ff7703af3056af502a695e1cd/manage.py#L350-L398) procedure in artsy/substance? I'd figure that we to minimize human intervention whenever possible.

**Comment by dleve123** [comment]
@izakp this is super useful, thanks for this. commented inline!

File: playbooks/kubernetes.md (hunk 1 above)
Comment (izakp) [reply playbooks/kubernetes.md]: Added a note for this as well

File: playbooks/kubernetes.md (hunk 2 above)
Comment (izakp) [reply playbooks/kubernetes.md]: > Also, IIRC, assuming > 1 pods for a service, K8s should distribute across many hosts, correct?

Yes this is the default scheduling for a Deployment exposed by a Service.  Will mention that.

File: playbooks/kubernetes.md (hunk 3 above)
Comment (izakp) [reply playbooks/kubernetes.md]: Yes this is temporary.  Will mention that.

**Comment by izakp** [comment]
Thanks for the detailed feedback @dleve123 - sorry I lost this thread but made your suggested changes and merging.

**Pull request merged**